src/pad.js

```diff
@@ -0,0 +1,109 @@
+        mappedLocation[i] = inputShape[i] - 1;
+      } else {
+        mappedLocation[i] -= beginningPadding[i];
+      }
+    }
+  } else {
+    // reflection mode or symmetric mode
+    const offset = mode === 'symmetric' ? 1 : 0;
+    for (let i = 0; i < length; i++) {
+      if (mappedLocation[i] < beginningPadding[i]) {
+        // mappedLocation[i] =
+        //    beginningPadding[i] + (beginningPadding[i] - mappedLocation[i]) -
+        //    beginningPadding[i] - offset;
+        mappedLocation[i] = beginningPadding[i] - mappedLocation[i] - offset;
+      } else if (mappedLocation[i] >= beginningPadding[i] + inputShape[i]) {
+        // mappedLocation[i] =
+        //    beginningPadding[i] + inputShape[i] - 1 -
+        //    (mappedLocation[i] - (beginningPadding[i] + inputShape[i] -1)) -
+        //    beginningPadding[i] + offset;
+        mappedLocation[i] =
+            beginningPadding[i] + inputShape[i] * 2 - mappedLocation[i] - 2 + offset;
+      } else {
+        mappedLocation[i] -= beginningPadding[i];
+      }
+    }
+  }
+  return mappedLocation;
+}
+
+/**
+ * Update element value of destination tensor.
+ * @param {Number} index
+ * @param {Tensor} source
+ * @param {Tensor} destination
+ * @param {Array} beginningPadding
+ * @param {MLPaddingMode} mode
+ * @param {Number} value
+ * @return {Tensor}
+ */
+function updateOutput(index, source, destination, beginningPadding, mode, value) {
+  const inputShape = source.shape;
+  const location = destination.locationFromIndex(index);
+  const length = location.length;
+  let bFillPadding = false;
+  for (let j = 0; j < length; j++) {
+    if (location[j] < beginningPadding[j] || location[j] >= beginningPadding[j] + inputShape[j]) {
+      bFillPadding = true;
+      break;
+    }
+  }
+  let result;
+  if (bFillPadding) {
+    if (mode === 'constant') {
+      result = value;
+    } else if (mode === 'edge' || mode === 'reflection' || mode === 'symmetric') {
+      const targetLocation = getMappedLocation(location, inputShape, beginningPadding, mode);
+      result = source.getValueByLocation(targetLocation);
+    }
+  } else {
+    const inputLocation = location.map((v, d) => v - beginningPadding[d]);
+    result = source.getValueByLocation(inputLocation);
+  }
+  destination.setValueByIndex(index, result);
+  return destination;
+}
+
+/**
+ * Inflate the tensor with constant or mirrored values on the edges.
+ * @param {Tensor} input
+ * @param {Array} beginningPadding
+ * @param {Array} endingPadding
+ * @param {MLPadOptions} [options]
+ * @return {Tensor}
+ */
+export function pad(
+    input,
+    beginningPadding,
+    endingPadding,
+    {
+      mode='constant',
+      value=0,
+    } = {}) {
+  const outputShape = input.shape.map((v, i) => v + beginningPadding[i] + endingPadding[i]);
+  let output = new Tensor(outputShape);
+  for (let i = 0; i < output.size; ++i) {
+    output = updateOutput(i, input, output, beginningPadding, mode, value);
```

Comment:
Is this correct to overwrite `output`? Doesn't `updateOutput` already call `setValueByIndex`? Otherwise line 108 will only return the last loop value.

Reply:
Yes, this **for** loop is to set each element of output tensor by index.
Maybe the function name `updateOutput`  confuse you, how about renaming it as `updateElement` or something, any suggestions? Thanks.

Reply:
(I missed this notification, and the CR is already completed meaning it doesn't matter much now, but to answer your question)

Couldn't this:

```
for (let i = 0; i < output.size; ++i) {
    output = updateOutput(i, input, output, beginningPadding, mode, value);
}
```

Have been this?...

```
for (let i = 0; i < output.size; ++i) {
    updateOutput(i, input, output, beginningPadding, mode, value);
}
```

...since `updateOutput` is already calling `setValueByIndex` on the destination.